sync2/txnid.go

```diff
@@ -1,38 +1,98 @@
+//     list. If the device ID list is now empty, the poller emits an "all clear"
+//     V2TransactionID payload.
+//
+// This is a best-effort affair to ensure that the rest of the proxy can wait for
+// transaction IDs to appear before transmitting an event down /sync to its sender.
+//
+// It's possible that we add an entry to this map and then the list of remaining
+// device IDs becomes out of date, either due to a new device creation or an
+// existing device expiring. We choose not to handle this case, because it is relatively
+// rare.
+//
+// To avoid the map growing without bound, we use a ttlcache and drop entries
+// after a short period of time.
+type PendingTransactionIDs struct {
+	// mu guards the pending field.
+	mu      sync.Mutex
+	pending *ttlcache.Cache
+	// loader should provide the list of device IDs
+	loader loaderFunc
 }
 
-func NewTransactionIDCache() *TransactionIDCache {
+func NewPendingTransactionIDs(loader loaderFunc) *PendingTransactionIDs {
 	c := ttlcache.NewCache()
 	c.SetTTL(5 * time.Minute)     // keep transaction IDs for 5 minutes before forgetting about them
 	c.SkipTTLExtensionOnHit(true) // we don't care how many times they ask for the item, 5min is the limit.
-	return &TransactionIDCache{
-		cache: c,
+	return &PendingTransactionIDs{
+		mu:      sync.Mutex{},
+		pending: c,
+		loader:  loader,
 	}
 }
 
-// Store a new transaction ID received via v2 /sync
-func (c *TransactionIDCache) Store(userID, eventID, txnID string) {
-	c.cache.Set(cacheKey(userID, eventID), txnID)
-}
+// MissingTxnID should be called to report that this device ID did not see a
+// transaction ID for this event ID. Returns true if this is the first time we know
+// for sure that we'll never see a txn ID for this event.
+func (c *PendingTransactionIDs) MissingTxnID(eventID, userID, myDeviceID string) (bool, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 
-// Get a transaction ID previously stored.
-func (c *TransactionIDCache) Get(userID, eventID string) string {
-	val, _ := c.cache.Get(cacheKey(userID, eventID))
-	if val != nil {
-		return val.(string)
+	data, err := c.pending.Get(eventID)
+	if err == ttlcache.ErrNotFound {
+		data = c.loader(userID)
+	} else if err != nil {
+		return false, fmt.Errorf("PendingTransactionIDs: failed to get device ids: %w", err)
 	}
-	return ""
+
+	deviceIDs, ok := data.([]string)
+	if !ok {
+		return false, fmt.Errorf("PendingTransactionIDs: failed to cast device IDs")
+	}
+
+	deviceIDs, changed := removeDevice(myDeviceID, deviceIDs)
+	if changed {
+		err = c.pending.Set(eventID, deviceIDs)
+		if err != nil {
+			return false, fmt.Errorf("PendingTransactionIDs: failed to set device IDs: %w", err)
+		}
+	}
+	return changed && len(deviceIDs) == 0, nil
+}
+
+// SeenTxnID should be called to report that this device saw a transaction ID
+// for this event.
+func (c *PendingTransactionIDs) SeenTxnID(eventID string) error {
+	c.mu.Lock()
+	c.mu.Unlock()
```

Comment:
No-op due to lack of defer, plus ttlcache is thread safe iirc

Reply:
Thanks.

ttlcache is threadsafe; that's not what the mutex is for. I think we need a Mutex here to ensure that the Get+Set in MissingTxnID is atomic. For example:

- We have three pollers A, B, C.
- Poller A sees an event without txn id and calls MissingTxnID.
  - `c.pending.Get()` fails, so we load up all device IDs: [A, B, C]
  - then `c.pending.Set()` with [B, C]
- Poller B sees the same event, also missing txn ID and calls MissingTxnID.
- Poller C does the same concurrently.

If the Get+Set isn't atomic, then we might do e.g.
- B gets [B, C] and prepares to write [C].
- C gets [B, C] and prepares to write [B].
- Last writer wins. Either way, we never write [] and so never return the all-clear signal. That would mean the API process hits its maximum grace period before returning the event to its sender.

Having written all that, the above wouldn't be the end of the world---particularly if this race is unlikely. But when I was thinking this through I was erring towards caution.

Reply:
Yup SGTM